readme.md
# bcp-47 [![Build Status][travis-badge]][travis] [![Coverage Status][codecov-badge]][codecov]

Parse and stringify [BCP 47][spec] language tags.

## Installation

[npm][npm-install]:

```bash
npm install bcp-47
```

## Table of Contents

*   [Usage](#usage)
*   [API](#api)
    *   [bcp47.parse(tag\[, options\])](#bcp47parsetag-options)
    *   [bcp47.stringify(schema)](#bcp47stringifyschema)
    *   [schema](#schema)
    *   [function warning(reason, code, offset)](#function-warningreason-code-offset)
*   [License](#license)

## Usage

Dependencies:

```javascript
var bcp47 = require('bcp-47');
```

Parsing:

```javascript
var schema = bcp47.parse('hy-Latn-IT-arevela');
```

Yields:

```json
{
  "language": "hy",
  "extendedLanguageSubtags": [],
  "script": "Latn",
  "region": "IT",
  "variants": [
    "arevela"
  ],
  "extensions": [],
  "privateuse": [],
  "irregular": null,
  "regular": null
}
```

Compiling:

```javascript
var tag = bcp47.stringify(schema);
```

Yields:

```txt
hy-Latn-IT-arevela
```

## API

### `bcp47.parse(tag[, options])`

Parse a BCP 47 tag into a language schema.
Note that the algorithm is case-insensitive.

###### `options.normalize`

Whether to normalize legacy tags when possible (`boolean`, default:
`true`).  For example, `i-klingon` does not match the BCP 47 language
algorithm but is valid nonetheless.  It is suggested to use `tlh`
instead (the ISO 639-3 code for Klingon).  When `normalize` is true,
passing `i-klingon`, or other deprecated tags, is handled as if their
suggested valid tag was given.

###### `options.forgiving`

By default, when an error is encountered, an empty object is returned.
When in forgiving mode, all found values up to the point of the error
are included (`boolean`, default: `false`).  So, for example, where by
default `en-GB-abcdefghi` an empty object is returned (as the language
variant is too long), in `forgiving` mode the `language` of `schema` is
populated with `en` and the `region` is populated with `GB`.

###### `options.warning`

When given, `warning` is invoked when an error is encountered
([`Function`][warning]).

###### Returns

[`Schema`][schema].

###### Throws

When `tag` is `null` or `undefined`.

### `bcp47.stringify(schema)`

Compile a [`schema`][schema] into a BCP 47 language tag.

###### Returns

`string` — BCP 47 language tag.

### `schema`

A schema may have the following properties.  A schema is deemed empty
when it has neither `language`, `irregular`, `regular`, nor `privateuse`
(where an empty `privateuse` array is handled as no `privateuse`, too).

###### `schema.language`

Two or three character [ISO 639][iso-639] language code, four character
reserved language code, or 5 to 8 (inclusive) characters registered
language subtag (`string`).  For example, `en` (English) or `cmn`
(Mandarin Chinese).

###### `schema.extendedLanguageSubtags`

Selected three-character [ISO 639][iso-639] codes(`Array.<string>`),
such as `yue` in `zh-yue-HK` (Chinese, Cantonese, as used in Hong Kong
SAR).

###### `schema.script`

Four character [ISO 15924][iso-15924] script code (`string`), such as
`Latn` in `hy-Latn-IT-arevela` (Eastern Armenian written in Latin
script, as used in Italy).

###### `schema.region`

Two alphabetical character [ISO 3166-1][iso-3166-1] code, or three
digit [UN M.49][un-m49] code (`string`).  For example, `CN` in
`cmn-Hans-CN` (Mandarin Chinese, Simplified script, as used in China).

###### `schema.variants`

5 to 8 (inclusive) character language variants (`Array.<string>`), such
as both `rozaj` and `biske` in `sl-rozaj-biske` (San Giorgio dialect
of Resian dialect of Slovenian).

###### `schema.extensions`

List of extensions (`Array.<Object>`), each an object containing a one
character `singleton`, and a list of `extensions` (`string`).
`singleton` cannot be `x` (case-insensitive), and `extensions` must be
between two and eight (inclusive) characters.  For example, an extension
would be `u-co-phonebk` in `de-DE-u-co-phonebk` (German, as used in
Germany, using German phonebook sort order), where `u` is the `singleton`
and `co` and `phonebk` are its extensions.

###### `schema.privateuse`

List of private-use subtags (`Array.<string>`), where each subtag must
be between one and eight (inclusive) characters.

###### `schema.regular`

One of the `regular` tags (`string`): tags which are seen as something
different by the algorithm.

Valid values are:

*   `art-lojban`;
*   `cel-gaulish`;
*   `no-bok`;
*   `no-nyn`;
*   `zh-guoyu`;
*   `zh-hakka`;
*   `zh-min`;
*   `zh-min-nan`;
*   `zh-xiang`.

###### `schema.irregular`

One of the `irregular` tags (`string`): tags which are seen as
invalid by the algorithm).

Valid values are:

*   `en-GB-oed`;
*   `i-ami`;
*   `i-bnn`;
*   `i-default`;
*   `i-enochian`;
*   `i-hak`;
*   `i-klingon`;
*   `i-lux`;
*   `i-mingo`;
*   `i-navajo`;
*   `i-pwn`;
*   `i-tao`;
*   `i-tay`;
*   `i-tsu`;
*   `sgn-BE-FR`;
*   `sgn-BE-NL`;
*   `sgn-CH-DE`.

### `function warning(reason, code, offset)`

Invoked when an error occurs.

###### Parameters

*   `reason` (`string`) — English reason for failure;
*   `code` (`number`) — Code for failure;
*   `offset` (`number`) — Index-based position of error.

###### Warnings

| code | reason                                                                 |
| :--- | :--------------------------------------------------------------------- |
| 1    | Too long variant, expected at most 8 characters                        |
| 2    | Too long extension, expected at most 8 characters                      |
| 3    | Too many extended language subtags, expected at most 3 subtags         |
| 4    | Empty extension, extensions must have at least 2 characters of content |
| 5    | Too long private-use area, expected at most 8 characters               |
| 6    | Found superfluous content after tag                                    |

## License

[MIT][license] © [Titus Wormer][author]

<!-- Definitions -->

[travis-badge]: https://img.shields.io/travis/wooorm/bcp-47.svg

[travis]: https://travis-ci.org/wooorm/bcp-47

[codecov-badge]: https://img.shields.io/codecov/c/github/wooorm/bcp-47.svg

[codecov]: https://codecov.io/github/wooorm/bcp-47

[npm-install]: https://docs.npmjs.com/cli/install

[license]: LICENSE

[author]: http://wooorm.com

[spec]: https://tools.ietf.org/html/bcp47

[warning]: #function-warningreason-code-offset

[schema]: #schema

[iso-639]: https://en.wikipedia.org/wiki/ISO_639

[iso-15924]: https://en.wikipedia.org/wiki/ISO_15924

[iso-3166-1]: https://en.wikipedia.org/wiki/ISO_3166-1

[un-m49]: https://en.wikipedia.org/wiki/UN_M.49


history.md
<!--remark setext-->

<!--lint disable no-multiple-toplevel-headings heading-style-->

1.0.2 / 2016-07-23
==================

*   Add missing dev-dependency ([`303073d`](https://github.com/wooorm/bcp-47/commit/303073d))

1.0.1 / 2016-07-23
==================

*   Refactor code-style to use `xo` ([`c72668e`](https://github.com/wooorm/bcp-47/commit/c72668e))

1.0.0 / 2016-02-21
==================


test/index.js
/**
 * @author Titus Wormer
 * @copyright 2016 Titus Wormer
 * @license MIT
 * @module bcp-47
 * @fileoverview Test suite for `bcp-47`.
 */

'use strict';

/* Tests. */
require('./parse');
require('./stringify');
require('./fixtures');


test/parse.js
/**
 * @author Titus Wormer
 * @copyright 2016 Titus Wormer
 * @license MIT
 * @module bcp-47
 * @fileoverview Test suite for `bcp-47`.
 */

'use strict';

/* Dependencies. */
var test = require('tape');
var bcp47 = require('..');

/* Tests. */
test('.parse()', function (t) {
  t.equal(typeof bcp47.parse, 'function', 'should be a method');

  t.throws(function () {
    bcp47.parse();
  }, 'should throw when given `undefined`');

  t.throws(function () {
    bcp47.parse(null);
  }, 'should throw when given `null`');

  t.doesNotThrow(function () {
    bcp47.parse({
      /** Stringify. */
      toString: function () {
        return 'en';
      }
    });
  }, 'should coerce to a string');

  t.deepEqual(
    bcp47.parse('i-klingon'),
    {
      language: 'tlh',
      extendedLanguageSubtags: [],
      script: null,
      region: null,
      variants: [],
      extensions: [],
      privateuse: [],
      irregular: null,
      regular: null
    },
    'should normalize when possible'
  );

  t.deepEqual(
    bcp47.parse('i-klingon', {normalize: false}),
    {
      language: null,
      extendedLanguageSubtags: [],
      script: null,
      region: null,
      variants: [],
      extensions: [],
      privateuse: [],
      irregular: 'i-klingon',
      regular: null
    },
    'should not normalize when `normalize: false`'
  );

  t.deepEqual(
    bcp47.parse('i-default'),
    {
      language: null,
      extendedLanguageSubtags: [],
      script: null,
      region: null,
      variants: [],
      extensions: [],
      privateuse: [],
      irregular: 'i-default',
      regular: null
    },
    'should return an irregular when not normalizable'
  );

  t.deepEqual(
    bcp47.parse('zh-min'),
    {
      language: null,
      extendedLanguageSubtags: [],
      script: null,
      region: null,
      variants: [],
      extensions: [],
      privateuse: [],
      irregular: null,
      regular: 'zh-min'
    },
    'should return a regular when not normalizable'
  );

  t.test('Too long variant', function (st) {
    var fixture = 'en-GB-abcdefghi';

    st.plan(6);

    /**
     * Handle the warnings.
     */
    function warning() {
      st.equal(
        arguments[0],
        'Too long variant, expected at most 8 characters'
      );

      st.equal(arguments[1], 1);
      st.equal(arguments[2], 15);
      st.equal(arguments.length, 3);
    }

    st.deepEqual(
      bcp47.parse(fixture, {warning: warning}),
      {
        language: null,
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return `null`'
    );

    st.deepEqual(
      bcp47.parse(fixture, {forgiving: true}),
      {
        language: 'en',
        extendedLanguageSubtags: [],
        script: null,
        region: 'GB',
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return untill the error when `forgiving: true`'
    );

    st.end();
  });

  t.test('Too many subtags', function (st) {
    var fixture = 'aa-bbb-ccc-ddd-eee';

    st.plan(6);

    /**
     * Handle the warnings.
     */
    function warning() {
      st.equal(
        arguments[0],
        'Too many extended language subtags, ' +
        'expected at most 3 subtags'
      );

      st.equal(arguments[1], 3);
      st.equal(arguments[2], 14);
      st.equal(arguments.length, 3);
    }

    st.deepEqual(
      bcp47.parse(fixture, {warning: warning}),
      {
        language: null,
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return `null`'
    );

    st.deepEqual(
      bcp47.parse('aa-bbb-ccc-ddd-eee', {forgiving: true}),
      {
        language: 'aa',
        extendedLanguageSubtags: [
          'bbb',
          'ccc',
          'ddd'
        ],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return untill the error when `forgiving: true`'
    );
  });

  t.test('Too long extension', function (st) {
    var fixture = 'en-i-abcdefghi';

    st.plan(6);

    /**
     * Handle the warnings.
     */
    function warning() {
      st.equal(
        arguments[0],
        'Too long extension, expected at most 8 characters'
      );

      st.equal(arguments[1], 2);
      st.equal(arguments[2], 13);
      st.equal(arguments.length, 3);
    }

    st.deepEqual(
      bcp47.parse(fixture, {warning: warning}),
      {
        language: null,
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return `null`'
    );

    st.deepEqual(
      bcp47.parse(fixture, {forgiving: true}),
      {
        language: 'en',
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return untill the error when `forgiving: true`'
    );

    st.end();
  });

  t.test('Empty extension', function (st) {
    var fixture = 'en-i-a';

    st.plan(6);

    /**
     * Handle the warnings.
     */
    function warning() {
      st.equal(
        arguments[0],
        'Empty extension, extensions must have ' +
        'at least 2 characters of content'
      );

      st.equal(arguments[1], 4);
      st.equal(arguments[2], 4);
      st.equal(arguments.length, 3);
    }

    st.deepEqual(
      bcp47.parse(fixture, {warning: warning}),
      {
        language: null,
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return `null`'
    );

    st.deepEqual(
      bcp47.parse(fixture, {forgiving: true}),
      {
        language: 'en',
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return untill the error when `forgiving: true`'
    );

    st.end();
  });

  t.test('Too long private-use', function (st) {
    var fixture = 'en-x-abcdefghi';

    st.plan(6);

    /**
     * Handle the warnings.
     */
    function warning() {
      st.equal(
          arguments[0],
          'Too long private-use area, expected at most ' +
          '8 characters'
      );

      st.equal(arguments[1], 5);
      st.equal(arguments[2], 13);
      st.equal(arguments.length, 3);
    }

    st.deepEqual(
      bcp47.parse(fixture, {warning: warning}),
      {
        language: null,
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return `null`'
    );

    st.deepEqual(
      bcp47.parse(fixture, {forgiving: true}),
      {
        language: 'en',
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return untill the error when `forgiving: true`'
    );

    st.end();
  });

  t.test('Extra content', function (st) {
    var fixture = 'abcdefghijklmnopqrstuvwxyz';

    st.plan(6);

    /**
     * Handle the warnings.
     */
    function warning() {
      st.equal(
        arguments[0],
        'Found superfluous content after tag'
      );

      st.equal(arguments[1], 6);
      st.equal(arguments[2], 0);
      st.equal(arguments.length, 3);
    }

    st.deepEqual(
      bcp47.parse(fixture, {warning: warning}),
      {
        language: null,
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return `null`'
    );

    st.deepEqual(
      bcp47.parse(fixture, {forgiving: true}),
      {
        language: null,
        extendedLanguageSubtags: [],
        script: null,
        region: null,
        variants: [],
        extensions: [],
        privateuse: [],
        irregular: null,
        regular: null
      },
      'should return untill the error when `forgiving: true`'
    );

    st.end();
  });

  t.end();
});


test/fixtures.js
/**
 * @author Titus Wormer
 * @copyright 2016 Titus Wormer
 * @license MIT
 * @module bcp-47
 * @fileoverview Test suite for `bcp-47`.
 */

'use strict';

/* Dependencies. */
var fs = require('fs');
var path = require('path');
var test = require('tape');
var bcp47 = require('..');

/* Methods. */
var join = path.join;
var read = fs.readFileSync;
var dir = fs.readdirSync;

/* Fixtures. */
test('fixtures', function (t) {
  var base = join(__dirname, 'fixtures');

  dir(base)
    .filter(function (fileName) {
      return fileName.charAt(0) !== '.';
    })
    .forEach(function (fileName) {
      var filePath = join(base, fileName);
      var name = fileName.slice(0, fileName.indexOf('.'));
      var fixture = JSON.parse(read(filePath, 'utf8'));
      var schema = bcp47.parse(name, {normalize: false});
      var tag = bcp47.stringify(schema);

      t.test(name, function (st) {
        st.deepEqual(schema, fixture, 'should parse');
        st.deepEqual(tag, name, 'should stringify');
        st.end();
      });
    });

  t.end();
});


lib/stringify.js
/**
 * @author Titus Wormer
 * @copyright 2016 Titus Wormer
 * @license MIT
 * @module bcp-47:stringify
 * @fileoverview Stringify BCP 47 language tags.
 */

'use strict';

/* Expose. */
module.exports = stringify;

/* Characters. */
var C_DASH = '-';
var C_X_LOWER = 'x';

/**
 * Compile a language schema to a BCP 47 language tag.
 *
 * @param {Object} schema - Schema to compile.
 * @return {string} - BCP 47 language tag.
 */
function stringify(schema) {
  var result = [];
  var values;
  var index;
  var value;
  var length;

  if (schema && (schema.irregular || schema.regular)) {
    return schema.irregular || schema.regular;
  }

  if (schema && schema.language) {
    result = result.concat(
      schema.language,
      schema.extendedLanguageSubtags || [],
      schema.script || [],
      schema.region || [],
      schema.variants || []
    );

    values = schema.extensions || [];
    index = -1;
    length = values.length;

    while (++index < length) {
      value = values[index];

      if (
        value.singleton &&
        value.extensions &&
        value.extensions.length
      ) {
        result = result.concat(value.singleton, value.extensions);
      }
    }
  }

  if (schema && schema.privateuse && schema.privateuse.length) {
    result = result.concat(C_X_LOWER, schema.privateuse);
  }

  return result.join(C_DASH);
}


lib/parse.js
/**
 * @author Titus Wormer
 * @copyright 2016 Titus Wormer
 * @license MIT
 * @module bcp-47:parse
 * @fileoverview Parse BCP 47 language tags.
 */

'use strict';

/* Dependencies. */
var has = require('has');
var alphanumeric = require('is-alphanumerical');
var alphabetical = require('is-alphabetical');
var decimal = require('is-decimal');
var regular = require('./regular');
var normal = require('./normalize');

/* Expose. */
module.exports = parse;

/* Characters. */
var C_DASH = '-';
var C_X_LOWER = 'x';

/* Character codes. */
var CC_DASH = C_DASH.charCodeAt(0);
var CC_X_LOWER = C_X_LOWER.charCodeAt(0);

/* Constants. */
var MAX_SUBTAG = 8;
var MAX_RESERVED = 4;
var MAX_ISO_639 = 3;
var MIN_ISO_639 = 2;
var MAX_EXTENDED_LANGUAGE_SUBTAG_COUNT = 3;
var MAX_VARIANT = 8;
var MIN_ALPHANUMERIC_VARIANT = 5;
var MIN_VARIANT = 4;
var MAX_EXTENSION = 8;
var MAX_PRIVATE_USE = 8;

/* Error codes. */
var ERR_VARIANT_TOO_LONG = 1;
var ERR_EXTENSION_TOO_LONG = 2;
var ERR_TOO_MANY_SUBTAGS = 3;
var ERR_EMPTY_EXTENSION = 4;
var ERR_PRIVATE_USE_TOO_LONG = 5;
var ERR_EXTRA_CONTENT = 6;

/* Error codes to messages. */
var ERRORS = {};

ERRORS[ERR_VARIANT_TOO_LONG] = 'Too long variant, expected at most 8 ' +
  'characters';

ERRORS[ERR_EXTENSION_TOO_LONG] = 'Too long extension, expected at most ' +
  '8 characters';

ERRORS[ERR_TOO_MANY_SUBTAGS] = 'Too many extended language subtags, ' +
  'expected at most 3 subtags';

ERRORS[ERR_EMPTY_EXTENSION] = 'Empty extension, extensions must have ' +
  'at least 2 characters of content';

ERRORS[ERR_PRIVATE_USE_TOO_LONG] = 'Too long private-use area, ' +
  'expected at most 8 characters';

ERRORS[ERR_EXTRA_CONTENT] = 'Found superfluous content after tag';

/**
 * Check if `code` is `x` (lower-case).
 *
 * @param {number} code - Code to check.
 * @return {boolean} - If `code` is `x`.
 */
function x(code) {
  return code === CC_X_LOWER;
}

/**
 * Check if `code` is `-`.
 *
 * @param {number} code - Code to check.
 * @return {boolean} - If `code` is `-`.
 */
function dash(code) {
  return code === CC_DASH;
}

/**
 * Create an empty results object.
 *
 * @return {Object} - Results object.
 */
function empty() {
  return {
    language: null,
    extendedLanguageSubtags: [],
    script: null,
    region: null,
    variants: [],
    extensions: [],
    privateuse: [],
    irregular: null,
    regular: null
  };
}

/**
 * Parse a BCP 47 language tag.
 *
 * @param {string} tag - Language tag to parse.
 * @param {Object} [options={}] - Configuration.
 * @param {number} [options.offset=0] - Offset.
 * @return {Object} - Results.
 * @throws {Error} - When `tag` is `null` or `undefined`.
 */
function parse(tag, options) {
  var settings = options || {};
  var warning = settings.warning;
  var forgiving = settings.forgiving;
  var normalize = settings.normalize == null ? true : settings.normalize;
  var result = {};
  var sensitive;
  var code;
  var index;
  var count;
  var groupCount;
  var offset;

  /* Check input. */
  if (tag == null) {
    throw new Error('Expected string, got `' + tag + '`');
  }

  /**
   * Get the character code at `pos`.
   */
  function char(pos) {
    return tag.charCodeAt(pos);
  }

  /**
   * Fail.
   */
  function fail(offset, code) {
    if (warning) {
      warning(ERRORS[code], code, offset);
    }

    return forgiving ? result : empty();
  }

  result = empty();

  /* Let’s start. */
  tag = String(tag);
  sensitive = tag;
  tag = tag.toLowerCase();

  if (has(normal, tag)) {
    if (normalize && normal[tag]) {
      return parse(normal[tag]);
    }

    result[regular.indexOf(tag) === -1 ? 'irregular' : 'regular'] = sensitive;

    return result;
  }

  index = 0;
  code = char(index);

  while (alphabetical(code)) {
    index++;
    code = char(index);
  }

  /* Parse a normal tag. */
  if (index < MIN_ISO_639 || index > MAX_SUBTAG) {
    index = 0;
  } else {
    if (index > MAX_RESERVED) {
      // lang(subtag):
      result.language = sensitive.slice(0, index);
    } else if (index > MAX_ISO_639) {
      // lang(reserved):
      result.language = sensitive.slice(0, index);
    } else {
      groupCount = 0;

      // lang(iso-639):
      result.language = sensitive.slice(0, index);

      while (
        dash(char(index)) &&
        alphabetical(char(index + 1)) &&
        alphabetical(char(index + 2)) &&
        alphabetical(char(index + 3)) &&
        !alphabetical(char(index + 4))
      ) {
        if (groupCount >= MAX_EXTENDED_LANGUAGE_SUBTAG_COUNT) {
          return fail(index, ERR_TOO_MANY_SUBTAGS);
        }

        // extended lang subtag
        result.extendedLanguageSubtags.push(
          sensitive.slice(index + 1, index + 4)
        );

        index += 4;
        groupCount++;
      }
    }

    if (
        dash(char(index)) &&
        alphabetical(char(index + 1)) &&
        alphabetical(char(index + 2)) &&
        alphabetical(char(index + 3)) &&
        alphabetical(char(index + 4)) &&
        !alphabetical(char(index + 5))
    ) {
      // script(iso-15924):
      result.script = sensitive.slice(index + 1, index + 5);

      index += 5;
    }

    if (dash(char(index))) {
      if (
        alphabetical(char(index + 1)) &&
        alphabetical(char(index + 2)) &&
        !alphabetical(char(index + 3))
      ) {
        // region(iso-3166-1):
        result.region = sensitive.slice(index + 1, index + 3);
        index += 3;
      } else if (
        decimal(char(index + 1)) &&
        decimal(char(index + 2)) &&
        decimal(char(index + 3)) &&
        !decimal(char(index + 4))
      ) {
        // region(un-m49):
        result.region = sensitive.slice(index + 1, index + 4);
        index += 4;
      }
    }

    while (dash(char(index))) {
      offset = index + 1;
      count = 0;

      while (alphanumeric(char(offset))) {
        offset++;
        count++;

        if (count > MAX_VARIANT) {
          return fail(offset, ERR_VARIANT_TOO_LONG);
        }
      }

      if (count >= MIN_ALPHANUMERIC_VARIANT) {
        // variant(long):
        result.variants.push(sensitive.slice(index + 1, offset));
        index = offset;
      } else if (
        decimal(char(index + 1)) &&
        count >= MIN_VARIANT
      ) {
        // variant(short):
        result.variants.push(sensitive.slice(index + 1, offset));
        index = offset;
      } else {
        break;
      }
    }

    while (dash(char(index))) {
      if (
        x(char(index + 1)) ||
        !alphanumeric(char(index + 1)) ||
        !dash(char(index + 2)) ||
        !alphanumeric(char(index + 3))
      ) {
        break;
      }

      offset = index + 2;
      groupCount = 0;

      while (
        dash(char(offset)) &&
        alphanumeric(char(offset + 1)) &&
        alphanumeric(char(offset + 2))
      ) {
        offset += 2;
        count = 2;
        groupCount++;

        while (alphanumeric(char(offset))) {
          if (count > MAX_EXTENSION) {
            return fail(offset, ERR_EXTENSION_TOO_LONG);
          }

          offset++;
          count++;
        }
      }

      if (!groupCount) {
        return fail(offset, ERR_EMPTY_EXTENSION);
      }

      // extension():
      result.extensions.push({
        singleton: sensitive.charAt(index + 1),
        extensions: sensitive.slice(index + 3, offset).split(C_DASH)
      });

      index = offset;
    }
  }

  if (
    (index === 0 && x(char(0))) ||
    (index !== 1 && dash(char(index)) && x(char(index + 1)))
  ) {
    offset = index = index ? index + 2 : 1;

    while (
      dash(char(offset)) &&
      alphanumeric(char(offset + 1))
    ) {
      offset += 2;
      count = 1;

      while (alphanumeric(char(offset))) {
        if (count >= MAX_PRIVATE_USE) {
          return fail(offset, ERR_PRIVATE_USE_TOO_LONG);
        }

        offset++;
        count++;
      }

      result.privateuse.push(sensitive.slice(index + 1, offset));
      index = offset;
    }
  }

  if (index !== tag.length) {
    return fail(index, ERR_EXTRA_CONTENT);
  }

  return result;
}
